Remove permalink reliance

jbutkus opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Some sites don't have permalinks enabled (e.g. local dev environments) and can't use the WC site API functionality.

The `rest_get_url_prefix` elsewhere is only used in permalink context (e.g. `get_rest_url` uses it conditionally), while `$wp->query_vars` are always populated, even in rewrite condition, thus making them more accessible.

### How to test the changes in this Pull Request:

1. Check out this branch on a local WooCommerce test site.
2. Make a request to plugin installation API endpoint with a valid Authentication header `curl -H "Authorization: Bearer abcd" -v https://local-wp-site.test/?rest_route=wccom-site/v1/installer` .
3. Verify the request is processed and not aborted.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enable WooCommerce Site API on installations not using permalink

jbutkus committed:
Remove permalink reliance

Some sites don't have permalinks enabled (e.g. local dev environments) and
can't use the WC site API functionality.

The `rest_get_url_prefix` elsewhere is only used in permalink context (e.g.
`get_rest_url` uses it conditionally), while `$wp->query_vars` are always
populated, even in rewrite condition, thus making them more accessible.

denho reviewed (changes requested):
@justas, on my test site it gives me notices:
```
Notice: Trying to get property 'query_vars' of non-object in /var/www/html/wp-content/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php on line 128

Notice: Undefined variable: rest_route in /var/www/html/wp-content/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php on line 130
```
It seems like `global $wp;` is not available here.

jbutkus committed:
Account for early start-up phase

Using `$wp` doesn't work in certain scenarios, where the entire system
didn't complete bootstrapping.
The raw query, after escaping the values, should sufficiently cater for
the need.

claudiosanches commented:
@denho do you mind to test it again?

denho commented:
> @denho do you mind to test it again?

Sure, @claudiosanches. Will start my tomorrow morning by checking this.

claudiosanches committed:
Build the REST API URL before checking for wccom-site

claudiosanches commented:
@jbutkus I made a small change to make sure that the REST API URL is being constructed correctly when we receive `$_REQUEST['rest_route']`.

denho commented on includes/wccom-site/class-wc-wccom-site.php:
```diff
@@ -124,13 +124,12 @@ protected static function get_authorization_header() {
-		if ( isset( $_REQUEST['rest_route'] ) ) {
-			$route = esc_url_raw( wp_unslash( $_REQUEST['rest_route'] ) );
+		if ( isset( $_REQUEST['rest_route'] ) ) { // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
+			$route = esc_url_raw( get_rest_url( null, wp_unslash( $_REQUEST['rest_route'] ) ) ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized, WordPress.Security.NonceVerification.Recommended
```
@claudiosanches, is calling `get_rest_url()` necessary here? I get 
```
Fatal error: Uncaught Error: Call to a member function using_index_permalinks() on null in /var/www/html/wp-includes/rest-api.php:358
```
Seems like `global $wp_rewrite;` is not available in `get_rest_url()`.

denho reviewed (changes requested):
Receive a fatal error when pinging the route with the GET parameter on my local test site. The error is triggered in `get_rest_url()` function. Left a comment on the code line with more details.

jbutkus committed:
Revert "Build the REST API URL before checking for wccom-site"

This reverts commit 0674b421ec25951ea127bc62a868244e8f81dc3d.

denho commented on includes/wccom-site/class-wc-wccom-site.php:
```diff
@@ -124,11 +124,17 @@ protected static function get_authorization_header() {
-		$rest_prefix = trailingslashit( rest_get_url_prefix() );
-		$request_uri = esc_url_raw( wp_unslash( $request_uri ) );
+		$route = '/';
+		$rest_prefix = '';
```
It seems like `$route` value is overwritten anyway, so probably don't need it here at all. 

And `$rest_prefix = '';` could probably be placed within the `if ( isset( $_REQUEST['rest_route'] ) )` condition, as it is also overwritten if this condition is not met.

claudiosanches replied:
@jbutkus the problemwith the current solution is that is generating a whole new URL from that query string, also doesn't create the correct URL.

`$route` is equal to `http://wccom-site/v1/installer`, and `$rest_prefix` equal to `wccom-site/`, so no need to `esc_url_raw` in this case, also it's just to compare, so no need to sanitize like that in this case.

denho reviewed (commented):
Works as expected. A few changes to polish the code would be nice.

@jbutkus, don't you think `phpcs:ignore` comments are useful?

claudiosanches committed:
Fixed PHPCS warnings and code clean up

claudiosanches commented:
I kept the original code, but with some tweaks, should work fine now without rely in any WP global.

denho reviewed (approved):
Works for me.
